cmd/transcriber/call/live_captions.go

```diff
@@ -0,0 +1,462 @@
+		// can finish it all in time, and it will never be able to recover. This happens especially when
+		// number of calls * threads per call > numCPUs. We need to be able to relieve the pressure.
+		if len(window) >= windowPressureLimitSamples {
+			window = window[:0]
+			prevWindowLen = 0
+			prevTranscribedPos = 0
+			if err := t.client.SendWs(wsEvMetric, public.MetricMsg{
+				SessionID:  ctx.sessionID,
+				MetricName: public.MetricLiveCaptionsWindowDropped,
+			}, false); err != nil {
+				slog.Error("processLiveCaptionsForTrack: error sending wsEvMetric MetricLiveCaptionsWindowDropped",
+					slog.String("err", err.Error()),
+					slog.String("trackID", ctx.trackID))
+			}
+			continue
+		}
+
+		prevAudioAt = time.Now()
+		prevWindowLen = len(window)
+
+		vadSegments, err := sd.Detect(window)
+		if err != nil {
+			slog.Error("processLiveCaptionsForTrack: vad failed", slog.String("err", err.Error()))
+			continue
+		}
+		if err := sd.Reset(); err != nil {
+			slog.Error("failed to reset speech detector",
+				slog.String("err", err.Error()),
+				slog.String("trackID", ctx.trackID))
+		}
+
+		if len(vadSegments) == 0 {
+			continue
+		}
+
+		// Prepare the vad segments and the audio for transcription.
+		segments := convertToSegmentSamples(vadSegments, len(window))
+		segments = removeShortSpeeches(segments)
+		cleaned := cleanAudio(window, segments)
+
+		// Before sending off data to be transcribed, check if new data is silence.
+		// If it is silence, don't send it off.
+		newDataIsSilence, windowFinished := checkSilence(segments, prevTranscribedPos)
+		if windowFinished {
+			window = window[:0]
+			prevTranscribedPos = 0
+			prevWindowLen = 0
+			continue
+		}
+		if newDataIsSilence {
+			continue
+		}
+
+		// Track our new position and send off data for transcription.
+		prevTranscribedPos = len(cleaned)
+		transcribedCh := make(chan string)
+		pkg := captionPackage{
+			pcm:   cleaned,
+			retCh: transcribedCh,
+		}
+		select {
+		case t.transcriberQueueCh <- pkg:
+			break
+		default:
+			if err := t.client.SendWs(wsEvMetric, public.MetricMsg{
+				SessionID:  ctx.sessionID,
+				MetricName: public.MetricLiveCaptionsTranscriberBufFull,
+			}, false); err != nil {
+				slog.Error("processLiveCaptionsForTrack: error sending wsEvMetric MetricTranscriberBufFull",
+					slog.String("err", err.Error()),
+					slog.String("trackID", ctx.trackID))
+			}
+			close(transcribedCh)
+		}
+
+		// While audio is being transcribed, we need to cut down the window if it's > maxWindowSize.
+		window, prevTranscribedPos = cutWindowToSize(ctx.trackID, window, segments, prevTranscribedPos)
+		prevWindowLen = len(window)
+
+		// Use a for loop and a select so that we can drop ticks waiting for the transcriber.
+	waitForTranscription:
+		for {
+			select {
+			case <-ticker.C:
+				slog.Debug("processLiveCaptionsForTrack: dropped a tick waiting for the transcriber",
+					slog.String("trackID", ctx.trackID))
```

Comment:
Could we add a `continue` here and avoid the loop label altogether?

Reply:
@cpoile It would be `continue` in the ticker case and a break in the for loop after the select block.

Reply:
@streamer45 Ah, I _was_ misunderstanding what you were getting at. Yep, that's better.